haicrypt/hcrypt_tx.c

```diff
@@ -55,21 +55,21 @@ int HaiCrypt_Tx_GetBuf(HaiCrypt_Handle hhc, size_t data_len, unsigned char **in_
 int HaiCrypt_Tx_ManageKeys(HaiCrypt_Handle hhc, void *out_p[], size_t out_len_p[], int maxout)
 {
 	hcrypt_Session *crypto = (hcrypt_Session *)hhc;
-	hcrypt_Ctx *ctx = NULL;
 	int nbout = 0;
 
 	if ((NULL == crypto)
-	||  (NULL == (ctx = crypto->ctx))
+	||  (NULL == crypto->ctx)
 	||  (NULL == out_p)
 	||  (NULL == out_len_p)) {
-		HCRYPT_LOG(LOG_ERR, "ManageKeys: invalid params: crypto=%p crypto->ctx=%p\n", crypto, ctx);
+		HCRYPT_LOG(LOG_ERR, "ManageKeys: invalid params: crypto=%p out_p=%p out_len_p=%p\n",
+                                                         crypto,   out_p,   out_len_p);
 		return(-1);
 	}
 
 	/* Manage Key Material (refresh, announce, decommission) */
 	hcryptCtx_Tx_ManageKM(crypto);
 
-	ASSERT(ctx->status == HCRYPT_CTX_S_ACTIVE);
+	ASSERT(crypto->ctx && crypto->ctx->status == HCRYPT_CTX_S_ACTIVE);
```

Comment:
It's not the same `crypto->ctx` as before the call to `hcryptCtx_Tx_ManageKM`. That's the whole point of this fix.

Reply:
Before #2616 it was
```c++
        hcryptCtx_Tx_ManageKM(crypto);
	ctx = crypto->ctx;
	if (NULL == ctx) {
		HCRYPT_LOG(LOG_ERR, "%s", "crypto context not defined\n");
		return(-1);
	}
	ASSERT(ctx->status == HCRYPT_CTX_S_ACTIVE);
```
After
```c++
        hcryptCtx_Tx_ManageKM(crypto);
	ASSERT(ctx->status == HCRYPT_CTX_S_ACTIVE);
```

So the question is if the function should still return `-1` if `crypto->ctx` becomes `NULL` after `hcryptCtx_Tx_ManageKM(crypto)`?

I.e.:
```c++
        hcryptCtx_Tx_ManageKM(crypto);
	if (NULL == crypto->ctx) {
		HCRYPT_LOG(LOG_ERR, "%s", "crypto context not defined\n");
		return(-1);
	}
	ASSERT(crypto->ctx->status == HCRYPT_CTX_S_ACTIVE);
```

Reply:
I'm not sure. The condition is the following: The `crypto->ctx` contains a pointer to one of two context objects. These objects point to one another in the `alt` field. After creating the `crypto` object, both should exist, one of them should be pointed to by `ctx` and the `hcryptCtx_Tx_ManageKM` call may change the value of the `ctx` pointer (to that pointed by `alt`), but is not expected to set this pointer to NULL.

This check may be an assertion and may result in error return. I personally think that assertion is useless because if the pointer is NULL the application will crash anyway, while with error return the worst that could happen is that the connection could break, as long as this error is properly handled.

```suggestion
	if (NULL == crypto->ctx) {
		HCRYPT_LOG(LOG_ERR, "%s", "crypto context NULL after ManageKM call\n");
		return(-1);
	}
	ASSERT(crypto->ctx->status == HCRYPT_CTX_S_ACTIVE);
```